beacon-chain/db/kv/slasher.go

```diff
@@ -197,6 +197,54 @@ func (s *Store) SaveSlasherChunks(
 	})
 }
 
+// CheckAndUpdateForSlashableProposals takes in a list of proposals and for each,
+// checks if there already exists a proposal at the same slot+validatorIndex combination. If so,
+// we append that existing proposal to a list. Otherwise, we save the proposal to disk, then append nil
+// to a list. We then return the list to the caller.
+func (s *Store) CheckAndUpdateForSlashableProposals(
+	ctx context.Context, proposals []*slashertypes.CompactBeaconBlock,
+) ([]*slashertypes.CompactBeaconBlock, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.CheckAndUpdateForSlashableProposals")
+	defer span.End()
+	existingProposals := make([]*slashertypes.CompactBeaconBlock, 0, len(proposals))
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(proposalRecordsBucket)
+		for _, proposal := range proposals {
+			key, err := keyForValidatorProposal(proposal)
+			if err != nil {
+				return err
+			}
+			existingSigningRoot := bkt.Get(key)
+			if existingSigningRoot != nil {
```

Comment:
Is it a double vote if the signing roots are the same?

Reply:
It is only a double proposal if the signing roots are different

Reply:
Right, so in that case this code is incorrect because it only checks if a root exists. It should rather verify if a *different* root exists.
```
if existingSigningRoot != nil && !bytes.Equal(existingSigningRoot, proposal.SigningRoot)
```